lib/db-utils.js

```diff
@@ -0,0 +1,248 @@
+                    street,
+                    city,
+                    zip,
+                },
+                signUpLink,
+                latitude,
+                longitude,
+            },
+        })
+    );
+}
+
+async function writeScraperRunByRefId({ refId, locationRefId, timestamp }) {
+    await faunaQuery(
+        q.Create(q.Ref(q.Collection("scraperRuns"), refId), {
+            data: {
+                locationRef: q.Ref(q.Collection("locations"), locationRefId),
+                timestamp,
+            },
+        })
+    );
+}
+
+async function writeAppointmentsByRefId({
+    refId,
+    scraperRunRefId,
+    date,
+    numberAvailable,
+    signUpLink,
+    extraData,
+}) {
+    await faunaQuery(
+        q.Create(q.Ref(q.Collection("appointments"), refId), {
+            data: {
+                scraperRunRef: q.Ref(
+                    q.Collection("scraperRuns"),
+                    scraperRunRefId
+                ),
+                date,
+                numberAvailable,
+                signUpLink,
+                extraData,
+            },
+        })
+    );
+}
+
+/*
+    Index-based queries. This allows us to search our indexes (scraperRunsByLocation, appointmentsByScraperRun).
+*/
+async function getScaperRunsByLocation(locationRefId) {
+    const scraperRuns = await faunaQuery(
+        q.Map(
+            q.Paginate(
+                q.Match(
+                    q.Index("scraperRunsByLocation"),
+                    q.Ref(q.Collection("locations"), locationRefId)
+                )
+            ),
+            q.Lambda((x) => q.Get(x))
+        )
+    );
+    // console.log(
+    //     `for locationRefId ${locationRefId} got response ${JSON.stringify(
+    //         scraperRuns
+    //     )}`
+    // );
+    return scraperRuns;
+}
+
+async function getAppointmentsByScraperRun(scraperRunRefId) {
+    const appointments = await faunaQuery(
+        q.Map(
+            q.Paginate(
+                q.Match(
+                    q.Index("appointmentsByScraperRun"),
+                    q.Ref(q.Collection("scraperRuns"), scraperRunRefId)
+                )
+            ),
+            q.Lambda((x) => q.Get(x))
+        )
+    );
+    // console.log(
+    //     `for scraperRunRefId ${scraperRunRefId} got response ${JSON.stringify(
+    //         appointments
+    //     )}`
+    // );
+    return appointments;
+}
+
+/* 
+    Utility that for one scraper output, writes to all the tables (locations, scraperRuns, and appointments).
+    We will call this many times from main.js.
+*/
+async function writeScrapedData({
+    name,
+    street,
+    city,
+    zip,
+    availability,
+    hasAvailability,
+    extraData,
+    timestamp,
+    latitude,
+    longitude,
+    signUpLink,
+}) {
+    const locationRefId = generateLocationId({ name, street, city, zip });
+    try {
+        await retrieveItemByRefId("locations", locationRefId);
+    } catch (e) {
+        if (e.message === "instance not found") {
```

Comment:
From https://docs.fauna.com/fauna/current/api/fql/functions/get?lang=javascript:
> If the document does not exist, a "document not found" error is returned. Use the exists predicate to avoid "document not found" errors.

IDK how much we care, but this seems to be the "right" way to do this?

Reply:
yeah this is probably worth fixing. I'll try using that predicate to check if the doc exists, instead of catching it when it doesn't

Reply:
fixed!